databases/orientdb/src/main/java/com/gentics/mesh/graphdb/cluster/TxCleanupTask.java

```diff
@@ -0,0 +1,139 @@
+package com.gentics.mesh.graphdb.cluster;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.gentics.madl.tx.AbstractTx;
+import com.gentics.mesh.etc.config.GraphStorageOptions;
+import com.gentics.mesh.etc.config.MeshOptions;
+import com.gentics.mesh.util.Tuple;
+import com.google.common.collect.ImmutableSet;
+
+import io.vertx.core.Handler;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+
+/**
+ * Task which terminates stalled or long running commit operations.
+ */
+@Singleton
+public class TxCleanupTask implements Handler<Long> {
+
+	private static final Logger log = LoggerFactory.getLogger(TxCleanupTask.class);
+
+	private static final Map<Thread, Long> registeredThreads = new ConcurrentHashMap<>();
+
+	private static final Set<ClassMethod> interruptedMethods = ImmutableSet.of(
+		new ClassMethod(AbstractTx.class, "commit"));
+
+	private GraphStorageOptions storageOptions;
+
+	@Inject
+	public TxCleanupTask(MeshOptions options) {
+		this.storageOptions = options.getStorageOptions();
+	}
+
+	@Override
+	public void handle(Long event) {
+		checkTransactions();
+	}
+
+	/**
+	 * Check whether the provided stack trace contains committing methods.
+	 * 
+	 * @param stackTrace
+	 * @return
+	 */
+	private boolean isCommitting(StackTraceElement[] stackTrace) {
+		return Stream.of(stackTrace)
+			.map(ClassMethod::of)
+			.anyMatch(interruptedMethods::contains);
+	}
+
+	/**
+	 * Check whether there are any transactions which exceed the set time limit.
+	 */
+	public void checkTransactions() {
+		if (log.isDebugEnabled()) {
+			log.debug("Checking {} transaction threads", registeredThreads.size());
+		}
+		List<Thread> toInterrupt = registeredThreads.entrySet()
+			.stream().filter(entry -> {
+				long now = System.currentTimeMillis();
+				long dur = now - entry.getValue();
+				long limit = storageOptions.getTxCommitTimeout();
+				boolean exceedsLimit = dur > limit;
+				if (exceedsLimit) {
+					log.warn("Thread {} exceeds time limit of {} with duration {}.", entry.getKey(), limit, dur);
+				}
+				return exceedsLimit;
+			}).map(entry -> {
+				Thread t = entry.getKey();
+				return Tuple.tuple(t, t.getStackTrace());
+			})
+			.filter(tuple -> isCommitting(tuple.v2()))
```

Comment:
This filter is not necessary anymore, because the threads are now registered manually anyway.

Either use this filter or the registering mechanism. I prefer the registering mechanism.

Reply:
I'll remove the filter